### PR TITLE
[V0][Bugfix] Fix parallel sampling performance regression when guided decoding is enabled

### PR DESCRIPTION
**Pull request by shadeMe**
`ParallelSampleSequenceGroup.add_request` has to copy the original `SamplingParams` instance as many times as the number of requested samples. This is currently done with a `copy.deepcopy` call, which is not advisable as the the `logits_processors` field could contain arbitrary Python objects with expensive-to-copy state. This happens to be the case with the current guided decoding logits processors, scaling linearly with the value of `SamplingParams.n` and introducing a bottleneck in the hot path.

A similar issue was previous identified, and `SamplingParams.clone` was introduced to workaround this issue - it attempts to call a `clone` function on each logits processor object, with the assumption that classes can implement this method to minimize the overhead by performing shallow copies when possible. However, not all existing logits processors implement this method. Nor does the `ParallelSampleSequenceGroup` class avail itself of the `SamplingParams.clone` method.

This commit introduces the following changes:
* Modify `ParallelSampleSequenceGroup.add_request` to call `SamplingParams.clone` instead of `copy.deepcopy`.
* Update the logits processors of the `guidance`, `outlines` and `xgrammar` backends to expose a `clone` method for the efficient copying of mutable state.

The `lm-format-enforcer` backend was left untouched as the logits processor implementation is external to vLLM.

## Benchmark

For text generation w/t Nvidia L4, Phi-1.5, `n=3` in an async setup, we see the `ParallelSampleSequenceGroup.add_request` call dominating the runtime during a 180 second profile (after warm-up/with in-flight requests) of the original code (anywhere b'ween 60%-86% of the total runtime depending on the backend). With the above changes, this is essentially eliminated (0.01%-0.6%). 

### Guidance
guidance-before
guidance-after

### Outlines
outlines-before
outlines-after

### Xgrammar
xgrammar-before
xgrammar-after

### CODE REVIEW

**Comment by chaunceyjiang** [comment]
@shadeMe Hi, sorry for the off-topic question—how did you generate this performance chart?

**Review by mgoin — APPROVED** [review]
Seems reasonable to me, but would like @russellb or @aarnphm to confirm before merge

File: vllm/model_executor/guided_decoding/outlines_logits_processors.py [hunk 1]
```diff
@@ -56,6 +56,12 @@ def __init__(self, guide: Guide, reasoner: Optional[ReasoningParser]):
         self._fsm_state: defaultdict[int, Union[int,
                                                 CFGState]] = defaultdict(int)
 
+    def clone(self) -> "BaseLogitsProcessor":
+        cloned = copy.copy(self)
+        cloned._guide = self._guide.copy()
+        cloned._fsm_state = copy.deepcopy(self._fsm_state)
+        return cloned
+
```
Comment (aarnphm) [inline vllm/model_executor/guided_decoding/outlines_logits_processors.py:64]: I would like to get https://github.com/vllm-project/vllm/pull/15975 in first before assigning this private attrs.

**Review by aarnphm — APPROVED** [review]
I'm good with this for structured outputs, and good to merge in guidance and xgrammar first before #15975.

iirc we will have to deepcopy the logit processors regardless if users use a custom logit processor? so essentially this change in `sequence.py` could potentially be breaking for users in V0 engine?

**Comment by shadeMe** [comment]
> iirc we will have to deepcopy the logit processors regardless if users use a custom logit processor? so essentially this change in sequence.py could potentially be breaking for users in V0 engine?

Breaking perhaps along the same lines as the [original PR](https://github.com/vllm-project/vllm/pull/3099/files) that introduced the `SamplingParams.clone` method - this PR just brings the parallel sampling code inline with its non-parallel counterpart. We could theoretically preserve the existing behaviour while excluding the structured outputs processors, but it would result in leaky abstractions.

**Comment by russellb** [comment]
merged from main to see if that gets CI passing

**Comment by shadeMe** [comment]
The CI failures appear to be unrelated AFAICT? The failing tests use the default `n=1` and do not use structured outputs.

**Comment by DarkLight1337** [comment]
Can you merge from main to fix the CI failures?

**Comment by DarkLight1337** [comment]
It appears that the samplers test failure on main is caused by this PR. PTAL https://buildkite.com/vllm/ci/builds/20641/steps?jid=0196fcb9-d7f7-4ff4-ad54-260dfc784dae
